### PR TITLE
Feature: Intel ARC GPU support with IPEX

### PR DESCRIPTION
**Pull request by Disty0**
# Pull request checklist

- [*] The PR has a proper title. Use [Semantic Commit Messages](https://seesparkbox.com/foundry/semantic_commit_messages). (No more branch-name title please)
- [*] Make sure you are requesting the right branch.
- [*] Make sure this is ready to be merged into the relevant branch. Please don't create a PR and let it hang for a few days.
- [*] Ensure all tests are passing.
- [*] Ensure linting is passing.

# PR type

- new feature

# Description

- Adds Intel ARC GPU support with IPEX on Linux / WSL.
- Patches GradScaler to work with Intel GPU and presents IPEX as CUDA.
- Added some workarounds for no FP64 and DDP support on the main pipeline.

- Intel didn't release a torchaudio wheel so we had to built it from source and requirements-ipex.txt downloads that build (Python 3.10 only).

# Screenshot

Screenshot from 2023-09-06 18-27-09
Screenshot from 2023-09-06 18-11-52
Screenshot from 2023-09-06 17-21-19
Screenshot from 2023-09-06 04-05-04

# Localhost url to test on

- http://localhost:7865

### CODE REVIEW

**Review by Tps-F — DISMISSED** [review]
Nice PR!
Seems to be mostly OK.
I'll fix a few code that bother me later.

**Comment by Tps-F** [comment]
@Disty0 You hijacked the torch module so have you tried infer?

**Comment by Disty0** [comment]
> 

These are confirmed to be working on my end:
Infer / Real Time Infer / Voice Splitting / Training


And this video i made as a test (and as a joke) took 25 minutes in total to make.
22 mins to train, 30 seconds to split vocal, 30 seconds to infer and another 2 minute to manually combine vocal and instrumental.
(Used a random 10 mins of Jordan Peterson motivational talking video to train.)
https://www.youtube.com/watch?v=FLzHa96SEWQ

And real time infer:
https://www.youtube.com/watch?v=9wIwhPnzgg4

Latency at default settings (+/- 5):
PM: 100 ms
Harvest: 115 ms
Crepe: 130 ms
RMVPE: 160 ms

File: infer-web.py [hunk 1]
```diff
@@ -18,6 +18,14 @@
 import gradio as gr
 import numpy as np
 import torch
+try:
+    import intel_extension_for_pytorch as ipex # pylint: disable=import-error, unused-import
+    if torch.xpu.is_available():
+        from infer.modules.ipex import ipex_init
+        ipex_init()
+        print("Using IPEX")
```
Comment (fumiama) [inline infer-web.py:26]: Consider using logger.info to print a log.

File: infer-web.py [hunk 2]
```diff
@@ -18,6 +18,14 @@
 import gradio as gr
 import numpy as np
 import torch
+try:
+    import intel_extension_for_pytorch as ipex # pylint: disable=import-error, unused-import
```
Comment (fumiama) [inline infer-web.py:22]: Maybe you can move the ipex check to config.py and pass the test result to a variable in Config class ex. `has_xpu`

File: infer/lib/infer_pack/models.py [hunk 3]
```diff
@@ -1156,7 +1156,10 @@ def forward(self, x):
         b, c, t = x.shape
         if t % self.period != 0:  # pad first
             n_pad = self.period - (t % self.period)
-            x = F.pad(x, (0, n_pad), "reflect")
+            if hasattr(torch, "xpu") and torch.xpu.is_available() and x.dtype == torch.bfloat16:
```
Comment (fumiama) [inline infer/lib/infer_pack/models.py:1159]: Same. can use `if config.has_xpu: ...`

File: infer/lib/rmvpe.py [hunk 4]
```diff
@@ -2,6 +2,13 @@
 
 import numpy as np
 import torch
+try:
```
Comment (fumiama) [inline infer/lib/rmvpe.py:5]: Same, move to config.

File: infer/modules/train/train.py [hunk 5]
```diff
@@ -17,6 +17,18 @@
 from random import randint, shuffle
 
 import torch
+try:
```
Comment (fumiama) [inline infer/modules/train/train.py:20]: same

File: infer/modules/vc/pipeline.py [hunk 6]
```diff
@@ -362,7 +362,7 @@ def pipeline(
             )
             pitch = pitch[:p_len]
             pitchf = pitchf[:p_len]
-            if self.device == "mps":
+            if self.device == "mps" or (hasattr(torch, "xpu") and torch.xpu.is_available()):
```
Comment (fumiama) [inline infer/modules/vc/pipeline.py:365]: Same, one time `(hasattr(torch, "xpu") and torch.xpu.is_available())` calculation, read cached result everywhere.

File: tools/rvc_for_realtime.py [hunk 7]
```diff
@@ -14,6 +14,14 @@
 import pyworld
 import scipy.signal as signal
 import torch
+try:
```
Comment (fumiama) [inline tools/rvc_for_realtime.py:17]: same

**Review by fumiama — CHANGES_REQUESTED** [review]
Good job with just some small issues.

File: infer-web.py (hunk 1 above)
Comment (Disty0) [reply infer-web.py:26]: Fair

File: infer-web.py (hunk 2 above)
Comment (Disty0) [reply infer-web.py:22]: I can but we still have to run these;
```
import intel_extension_for_pytorch as ipex
from infer.modules.ipex import ipex_init
ipex_init()
```
Same thing for the rest.

Comment (fumiama) [reply infer-web.py:22]: > I can replace try with if has_xpu but we still have to run these;
> 
> ```
> import intel_extension_for_pytorch as ipex
> from infer.modules.ipex import ipex_init
> ipex_init()
> ```
> 
> Same thing for the rest.

Evev if a py file import the config.py and config.py had executed those?

File: infer/lib/rmvpe.py (hunk 4 above)
Comment (Disty0) [reply infer/lib/rmvpe.py:5]: Removing this causes Torch not enabled with CUDA errors.

File: infer/modules/train/train.py (hunk 5 above)
Comment (Disty0) [reply infer/modules/train/train.py:20]: This imports GradScaler and autocast.

File: infer/modules/vc/pipeline.py (hunk 6 above)
Comment (Disty0) [reply infer/modules/vc/pipeline.py]: Done.

File: infer/lib/infer_pack/models.py (hunk 3 above)
Comment (Disty0) [reply infer/lib/infer_pack/models.py]: models.py doesn't import config so i added has_xpu variable below the main imports.

Comment (fumiama) [reply infer/lib/infer_pack/models.py]: OK.

File: infer/modules/train/train.py (hunk 5 above)
Comment (fumiama) [reply infer/modules/train/train.py:20]: OK.

File: infer/lib/rmvpe.py (hunk 4 above)
Comment (fumiama) [reply infer/lib/rmvpe.py:5]: OK.

**Review by fumiama — APPROVED** [review]

**Pull request merged**